atl-checker/src/algorithms/certain_zero/search_strategy/linear_optimize.rs

```diff
@@ -0,0 +1,505 @@
+use priority_queue::PriorityQueue;
+use std::cmp;
+use std::collections::HashMap;
+use std::option::Option::Some;
+use std::sync::Arc;
+use BinaryOpKind::{
+    Addition, And, Division, Equality, GreaterOrEqual, GreaterThan, Implication, Inequality,
+    LessOrEqual, LessThan, Multiplication, Or, Subtraction, Xor,
+};
+
+/// A SearchStrategyBuilder for building the LinearOptimizeSearch strategy.
+pub struct LinearOptimizeSearchBuilder {
+    pub game: IntermediateLcgs,
+}
+
+impl SearchStrategyBuilder<AtlVertex, LinearOptimizeSearch> for LinearOptimizeSearchBuilder {
+    fn build(&self) -> LinearOptimizeSearch {
+        LinearOptimizeSearch::new(self.game.clone())
+    }
+}
+
+/// Search strategy using ideas from linear programming to order next vertices,
+/// based on distance from the vertex to a region that borders the line between true/false in the formula.
+pub struct LinearOptimizeSearch {
+    /// Priority based on distance, lowest distance highest priority
+    queue: PriorityQueue<Edge<AtlVertex>, i32>,
+    game: IntermediateLcgs,
+    /// Maps the hash of a Phi and usize of state, to a result distance
+    result_cache: HashMap<(Arc<Phi>, usize), i32>,
+    /// Maps usize from proposition, to a hashmap mapping symbols to the ranges they need to be within,
+    /// to satisfy the proposition
+    proposition_cache: HashMap<usize, HashMap<SymbolIdentifier, Ranges>>,
+    /// Maps hash of a phi to a rangedphi
+    phi_cache: HashMap<Arc<Phi>, RangedPhi>,
+    // TODO - could add a new cache, to compare distance between states and use this to guesstimate distance
+    // TODO - to acceptance region, based on previous results (Mathias supervisor suggestion)
+}
+
+impl LinearOptimizeSearch {
+    pub fn new(game: IntermediateLcgs) -> LinearOptimizeSearch {
+        LinearOptimizeSearch {
+            queue: PriorityQueue::new(),
+            game,
+            result_cache: HashMap::new(),
+            proposition_cache: HashMap::new(),
+            phi_cache: HashMap::new(),
+        }
+    }
+}
+
+/// Holds the range for which a symbol should be within, to satisfy a proposition
+#[derive(Clone)]
+pub enum Ranges {
+    /// Must be the case that the state of the symbol is inside this range to satisfy the propositon (inclusive)
+    Range(i32, i32),
+}
+
+/// Used when mapping a Phi to a ranged variant,
+/// where the propositions have been replaced by hashmaps mapping symbols to Ranges
+pub enum RangedPhi {
+    /// Either or should hold
+    Or(Box<RangedPhi>, Box<RangedPhi>),
+    /// Both should hold
+    And(Box<RangedPhi>, Box<RangedPhi>),
+    /// Mapping symbols to ranges
+    Proposition(HashMap<SymbolIdentifier, Ranges>),
+    True,
+    False,
+}
+
+/// Holds extracted linear expressions from the formula in AtlVertex
+#[derive(Eq, PartialEq, Hash, Clone, Debug)]
+struct LinearExpression {
+    pub symbol: SymbolIdentifier,
+    pub constant: i32,
+    pub operation: BinaryOpKind,
```

Comment:
Replace BinaryOpKind with a new enum or use ComparisonOp, since you know that it will never be addition/subtraction/division/etc anyway. Only comparisons.